File: hedera-mirror-rest/controllers/bound.js
/*-
 * ‌
 * Hedera Mirror Node
 * ​
 * Copyright (C) 2019 - 2023 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

import _ from 'lodash';

import {InvalidArgumentError} from '../errors';
import * as utils from '../utils';

/**
 * Holds the filters for query parameters in a multi-column paging query. The parsing logic only allows single
 * occurrence of equal, lt/lte, and gt/gte filters, respectively.
 *
 * FilterKey may be referenced in validation errors.
 */
class Bound {
  constructor(filterKey, viewModelKey) {
    this.filterKey = filterKey;
    this.viewModelKey = !_.isNil(viewModelKey) ? viewModelKey : filterKey;
    this.equal = null;
    this.lower = null;
    this.upper = null;
  }

  hasBound() {
    return this.hasLower() || this.hasUpper();
  }

  hasEqual() {
    return !_.isNil(this.equal);
  }

  hasLower() {
    return !_.isNil(this.lower);
  }

  hasUpper() {
    return !_.isNil(this.upper);
  }

  isEmpty() {
    return !this.hasEqual() && !this.hasLower() && !this.hasUpper();
  }

  parse(filter) {
    const operator = filter.operator;
    if (operator === utils.opsMap.eq) {
      if (this.hasEqual()) {
        throw new InvalidArgumentError(`Only one equal (eq) operator is allowed for ${this.filterKey}`);
      }
      this.equal = filter;
    } else if (utils.gtGte.includes(operator)) {
      if (this.hasLower()) {
        throw new InvalidArgumentError(`Only one gt/gte operator is allowed for ${this.filterKey}`);
      }
      this.lower = filter;
    } else if (utils.ltLte.includes(operator)) {
      if (this.hasUpper()) {
        throw new InvalidArgumentError(`Only one lt/lte operator is allowed for ${this.filterKey}`);
      }
      this.upper = filter;
    } else {
      throw new InvalidArgumentError(`Not equal (ne) operator is not supported for ${this.filterKey}`);
    }
  }

  // for test only
  static create(properties) {
    return Object.assign(new Bound(), properties);
  }
}

export default Bound;

File: hedera-mirror-rest-java/src/main/java/com/hedera/mirror/restjava/service/Bound.java
/*
 * Copyright (C) 2024-2025 Hedera Hashgraph, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.hedera.mirror.restjava.service;

import static com.hedera.mirror.restjava.common.RangeOperator.EQ;
import static com.hedera.mirror.restjava.common.RangeOperator.GT;
import static com.hedera.mirror.restjava.common.RangeOperator.LT;

import com.hedera.mirror.restjava.common.NumberRangeParameter;
import com.hedera.mirror.restjava.common.RangeOperator;
import com.hedera.mirror.restjava.common.RangeParameter;
import java.util.Arrays;
import java.util.EnumMap;
import lombok.Getter;
import org.apache.commons.lang3.ArrayUtils;
import org.apache.commons.lang3.StringUtils;
import org.jooq.Field;

public class Bound {

    public static final Bound EMPTY = new Bound(null, false, StringUtils.EMPTY, null);

    @Getter
    private final Field<Long> field;

    @Getter
    private RangeParameter<Long> lower;

    @Getter
    private RangeParameter<Long> upper;

    private final String parameterName;

    private final EnumMap<RangeOperator, Integer> cardinality = new EnumMap<>(RangeOperator.class);

    public Bound(RangeParameter<Long>[] params, boolean primarySortField, String parameterName, Field<Long> field) {
        this.field = field;
        this.parameterName = parameterName;

        if (ArrayUtils.isEmpty(params)) {
            return;
        }

        for (var param : params) {
            if (param.hasLowerBound()) {
                lower = param;
            } else if (param.hasUpperBound()) {
                upper = param;
            }
            cardinality.merge(param.operator(), 1, Math::addExact);
        }

        long adjustedLower = getAdjustedLowerRangeValue();
        long adjustedUpper = adjustUpperBound();
        if (primarySortField && adjustedLower > adjustedUpper) {
            throw new IllegalArgumentException("Invalid range provided for %s".formatted(parameterName));
        }
    }

    public long adjustUpperBound() {
        if (this.upper == null) {
            return Long.MAX_VALUE;
        }

        long upperBound = this.upper.value();
        if (this.upper.operator() == RangeOperator.LT) {
            upperBound--;
        }

        return upperBound;
    }

    public RangeParameter<Long> adjustLowerRange() {
        if (this.hasEqualBounds()) {
            // If the primary param has a range with a single value, rewrite it to EQ
            lower = new NumberRangeParameter(EQ, this.getAdjustedLowerRangeValue());
            upper = null;
        }

        return lower;
    }

    public long getAdjustedLowerRangeValue() {
        if (this.lower == null) {
            return 0;
        }

        long lowerBound = this.lower.value();
        if (this.lower.operator() == RangeOperator.GT) {
            lowerBound++;
        }

        return lowerBound;
    }

    public void adjustUpperRange() {
        if (!this.isEmpty() && lower != null && lower.operator() == EQ) {
            // If the secondary param operator is EQ, set the secondary upper bound to the same
            upper = lower;
        }
    }

    // Gets a range value if the operator is converted from GT/LT to EQ/GTE/LTE
    public long getInclusiveRangeValue(boolean upper) {
        var rangeParameter = upper ? this.getUpper() : this.getLower();
        var operator = rangeParameter.operator();
        long value = rangeParameter.value();
        if (operator == GT) {
            value += 1L;
        } else if (operator == LT) {
            value -= 1L;
        }

        return value;
    }

    public int getCardinality(RangeOperator... operators) {
        return Arrays.stream(operators)
                .mapToInt(x -> cardinality.getOrDefault(x, 0))
                .sum();
    }

    public boolean isEmpty() {
        return lower == null && upper == null;
    }

    public boolean hasLowerAndUpper() {
        return lower != null && upper != null;
    }

    public boolean hasEqualBounds() {
        return hasLowerAndUpper() && getAdjustedLowerRangeValue() == adjustUpperBound();
    }

    // Returns a new bound with only a lower rangeParameter
    public Bound toLower() {
        return createBound(this.getLower());
    }

    // Returns a new bound with only an upper rangeParameter
    public Bound toUpper() {
        return createBound(this.getUpper());
    }

    public void verifyUnsupported(RangeOperator unsupportedOperator) {
        if (getCardinality(unsupportedOperator) > 0) {
            throw new IllegalArgumentException(
                    String.format("Unsupported range operator %s for %s", unsupportedOperator, parameterName));
        }
    }

    public void verifySingleOccurrence() {
        verifySingleOccurrence(RangeOperator.EQ);
        verifySingleOccurrence(RangeOperator.GT, RangeOperator.GTE);
        verifySingleOccurrence(RangeOperator.LT, RangeOperator.LTE);
    }

    public void verifyEqualOrRange() {
        if (this.getCardinality(RangeOperator.EQ) == 1
                && (this.getCardinality(RangeOperator.GT, RangeOperator.GTE) != 0
                        || this.getCardinality(RangeOperator.LT, RangeOperator.LTE) != 0)) {
            throw new IllegalArgumentException("Can't support both range and equal for %s".formatted(parameterName));
        }
    }

    private Bound createBound(RangeParameter<Long> param) {
        if (param == null) {
            return Bound.EMPTY;
        }

        var params = new NumberRangeParameter[] {new NumberRangeParameter(param.operator(), param.value())};
        return new Bound(params, false, parameterName, field);
    }

    private void verifySingleOccurrence(RangeOperator... rangeOperators) {
        if (this.getCardinality(rangeOperators) > 1) {
            throw new IllegalArgumentException(
                    "Only one range operator from %s is allowed for the given parameter for %s"
                            .formatted(Arrays.toString(rangeOperators), parameterName));
        }
    }
}
